### ISSUE TITLE
[RFC] Add pending locations to support easier onboarding

### ISSUE THREAD

**Issue by dtuite** [backstage/backstage#2801]
**Status:** Open for comments

## Need

Imagine you are implimenting Backstage at your comany. Your company has 200 GitHub repos which you want to appear in Backstage in some fashion. You need to somehow put a `catalog-info.yaml` file at the root of each repo.

You have a change cannon like [prbot](https://github.com/dsymonds/prbot) or [cannon](https://github.com/TouchBistro/cannon) at your disposal. You fire it up and automatically open a PR against each of the 200 repos. You pick the component name and description based on the repo name and default `kind` to `Component` just to get started.

You now have 200 open PRs. When a PR is merged, you get a GitHub notification and you must add it into Backstage so it shows up in the catalog. It may take weeks for some teams to merge the PR so you are tracking this work for a long time. 

You could script this perhaps but you might end up running the script for a long time or listening to GitHub webhooks or something.

It would be better if Backstage could track these "pending" components so there automatically appear in the catalog when the PR is merged.

## Proposal

Add a concept of "pending locations" to Backstage so that repos can be tracked before they have a `catalog-info.yaml` file. These locations should automatically promote themselves to "active locations" when the YAML file appears in a valid state.

## Alternatives

 1. This could be scripted in the contrib directory and ran outside of Backstage to assist the user.
 2. Something with GitHub actions or similar which pings Backstage when a `catalog-info.yaml` file is found.
 3. Support both YAML-based and database-only entities in Backstage. The user could then choose not to bother with the YAMLs if they desire.
 4. The user could add the newly created branches to the catalog as soon as the PR is open but the locations would then die if the branch was deleted after the PR was merged.

## Risks

Not sure.

**Comment by Rugvip** [comment]
Both implementation-wise and UX-wise, what would be the difference between a pending location, and a registered location pointing to a `catalog-info.yaml` that doesn't exist yet?

**Comment by dtuite** [comment]
@mcalus3 is planning to comment on the possible implementation at some point.

**UX wise**

 1. The edit button in the About card has nowhere to link to if a location is pending. Alternatively, it can link to the branch of the open PR which adds the catalog-info.yaml file, but then it will be linking to the wrong place when the PR is subsequently merged.
 2. As a user, I will be confused if some Backstage entitites are backed off to YAML files in the main repo branch and others aren't, but there is nothing in the interface to indicate which is which.
 3. Currently there are errors in the logs for missing locations, which suggests there is a problem which needs to be fixed. Pending locations would be expected, so I would expect to see no log errors about them.

**Comment by dtuite** [comment]
@Rugvip I meant to tag you in the above answer to your question.

**Comment by Rugvip** [comment]
@dtuite Oh would we be initially set the location of the `catalog-info.yaml` to the PR branch/fork, and then switch it once the PR is merged? Would imagine that could work out pretty nicely.

**Comment by dtuite** [comment]
@Rugvip Yeah I think the "automatic switching" is the main thing. You're saying... "I expect a catalog-info.yaml to show up in this location at some point in the future (ie. when the PR is merged)". This feature makes it easy to track all of the repos you are working to add to Backstage, even if they are not yet oficially added because the PR is still open.

**Comment by freben** [comment]
> Something with GitHub actions or similar which pings Backstage when a catalog-info.yaml file is found.

This is the one I'm partial to. I would like to avoid any intermediate / incomplete states. This could probably work great with a github apps integration, such that we in an event driven fashion can pick up on catalog files as they appear, across an entire org, without manual registration.

**Comment by mcalus3** [comment]
We think about introducing intermediate / incomplete states for two reasons:
1. To keep the information in the backstage database that a pull request for given catalog file has been opened and waiting for merge. This could be presented to users to remind them and prevent accidental trying to create the same component twice.
2. To allow fast onboarding, so the user who are trying out backstage in their organization would be able to configure it in one run, without need to wait for PR merge to see the first component.

@freben do you think we could achieve these goals without introducing new "incomplete" state of the entity?

**Comment by andrewthauer** [comment]
I'm not suggesting this is an ideal solution, but as a general thought, but placeholder component entities could be automatically generated via a catalog processor that reads a GitHub org. For instance, something like the following could be emitted for each repo:

```yaml
apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  name: repo-name
  description: unknown
  annotations:
    github.com/project-slug: org/repo-name
spec:
  type: unkown
  owner: unknown
  lifecycle: unknown
```

The `unknown` (or some other placeholder value) could be treated as a way to say the entity is incomplete. The UI could potentially use these placeholder values to indicate to the user this entity is just a placeholder. The one thing I'm unsure of is if the location type/target managing the entity could be easily replaced later once say an actual `catalog-info.yaml` file is created.

**Comment by mcalus3** [comment]
I was thinking about implementing this functionality in client app. I can see in `catalog-backend` code that upon a regular existing component creation, client app is sending a request with the yaml configuration from file in github repo and adds an annotation
```backstage.io/managed-by-location: <path-to-the-yaml-file>```

Would it be possible to instead send
```backstage.io/managed-by-location: pending```
And then handle this pending state in the user interface? If entity location is pending, instead of "About" card in the overview show the info that the entity location has yet to be set, maybe even display some icon on the catalog list rows to indicate this.

**Comment by freben** [comment]
Yes, that annotation is a readonly thing effectively; it's a marker left by the processing chains that tell where the data came from.

**Comment by freben** [comment]
I'll read up a bit on this thread tomorrow again, and gather my thoughts.

**Comment by dtuite** [comment]
We talked over the annotation driven implementation this morning and it seems promising. We'll update later with a more thorough idea of implementaition.

**Comment by dtuite** [comment]
Here's the best plan we can see right now. Props to @andrewthauer for the idea.

1. Build a plugin UI which fetches a list of repos from GitHub and allows the user to choose which ones they want to import into Backstage.
2. When the user chooses a repo, the following happens:
   1. We directly insert metadata for the repo into Backstage via the API.

**Comment by dtuite** [comment]
Here's the best plan we have so far. Props to @andrewthauer for the idea! ❤️  

1. Build a plugin UI which fetches a list of repos from GitHub and allows the user to choose which ones they want to import into Backstage. This will focus primarily on repos which the user has write access to.
2. When the user chooses a repo, the following happens:
   1. We automatically create a branch against the repo, add commit to add a `catalog-info.yaml` file and submit a PR.
   2. We simultaneously and directly insert metadata for the repo into Backstage via the API. This metadata contains an annotation (probably `backstage.io/pending-pull-request:` or something like that) which points to the PR.
   3. We add a webhook to the repo to ping Backstage when the PR is merged.
3. Add a widget to Backstage which can detect the `backstage.io/pending-pull-request:` annotation and indicate to the user whether the PR has been merged or not.
4. Add an endpoint to Backstage backend which receives the webhook even and creates a Location for the entitiy in question.
5. Add a plugin UI which allows the user to list entities which have outstanding PRs and follow up on them.

Does that sound about right @mcalus3 ?

I'm not suggesting that we add all of this functionality directly into Backstage as some users might not want it. We're just trying to figure out if there is a reasonable path towards a UX like this.

What do you think @iain-b / @martina-if ?

**Comment by andrewthauer** [comment]
One thing I think would be very useful in backstage core vs a plugin, would be the underlying utility library / helpers for querying the GH org. I can see this being very useful to expose an API, so other catalog processor authors / integrators can leverage.

**Comment by martina-if** [comment]
In the long term I also like more the idea of having a Github App that finds, validates and pushes entities to the backstage catalog, but I think this is also a good way to do a quick onboarding since it has the added value of not requiring org level permissions. This means anyone can run this and get started on Backstage.

**Comment by mbruggmann** [comment]
Regarding permissions, actually I think the opposite is true: As an owner of a repository, you are allowed to add a Github App to it on your own. But the proposed method only works if Backstage has org- or installation-wide access to all repositories. 

A middle ground could be to have a `Repository` type in the catalog, which we could create through the API when a repo gets imported. That would then inform Backstage that we should scan those repositories for `catalog-info.yaml` files. Would that be possible to implement?

**Comment by freben** [comment]
Or make it a Location kind which already exists, pointing to the repo, and make a processor that can read entire repos.

**Comment by andrewthauer** [comment]
Would it be possible to covert this special Location kind into another (e.g. github/api) after a `catalog-info.yaml` PR was merged? Would say an component entity with the name `foo` (no namespace) mapped to to a different `managed-by-location` if another location was used to process it?

**Comment by dtuite** [comment]
@mbruggmann 

> But the proposed method only works if Backstage has org- or installation-wide access to all repositories.

How so? I'm worried I'm missing something now. Which part requires the elevated permissions?

**Comment by dtuite** [comment]
@freben 

> Or make it a Location kind which already exists, pointing to the repo, and make a processor that can read entire repos.

Interesting thought. Playing this out, are you suggesting that...

 1. The processor would check all branches for a valid `catalog-info.yaml`
 2. The processor could tell which is the main branch of the repo (`master` in most cases, but not guaranteed)
 3. The processor could therefore tell if the entity was in a "pending" state (ie. the PR has not yet been merged).

**Comment by freben** [comment]
As per the original post:

> automatically appear in the catalog when the PR is merged.

I think this is the part we should focus on. I don't think we should (at least at this point) think of _entities_ as being in a pending state, therefore, there would be no reason to look at non-main branches. The value proposition that I agree with is in automatically picking up the entity _after merge_ without any further user action, which a repo-targeted location will. Then it will at first be just like any ol' location, nothing special about it, besides it resulting in zero entities in its output because the PR hasn't landed yet.

Dealing with non-merged data results in a host of gnarly questions I think. What happens with collisions with entities elsewhere? With entities in the same location? With renames? With errors? Etc etc. I can't really wrap my head around that.

**Comment by dtuite** [comment]
Got it. Appreciate the explanation!

**Comment by mbruggmann** [comment]
+1 to everything @freben suggested.

@dtuite I might have misunderstood how this works, but what I meant with the permissions is that Backstage needs a token in the config file (eg here https://github.com/spotify/backstage/blob/master/app-config.yaml#L66) that can read from all repositories where we have registered components, and also all new repositories that we want to register components from.

**Comment by mcalus3** [comment]
@mbruggmann @dtuite, claryfing up the permissions thing - We are having the solution in which user logs in with their own GitHub account (through OAuth). Thanks to that, _backstage app GitHub account_ doesn't need any special permissions to the repos, pull requests are opened from the user's account (which is nice because repo owners would have the real person to talk to in the comments) and if the user doesn't have the _owner permission_ to create a webhook, he could ask someone who has (some sort of organization administrator) to complete a step that creates a webhook for him.

**Comment by freben** [comment]
@dtuite I believe this should be closed now that #3008 and the follow-up PRs are merged, correct?
